Q1. Write a Python function that takes in two arrays of data and calculates the F-value for a variance ratio 
test. The function should return the F-value and the corresponding p-value for the test
Ans. 

In [7]:
import numpy as np
from scipy.stats import f

array1=[17,19,23,24,28,29,37,45,51,55]
array2=[14,15,15,17.18,22,25,25,27,34]

def variance_ratio_test(array1,array2):
    var1=np.var(array1,ddof=1)
    var2=np.var(array2,ddof=1)
    
    df1=len(array1)-1
    df2=len(array2)-1
    
    f_value=var1/var2
    
    p_value=1-f.cdf(f_value,df1,df2)
    
    
    
    return f_value, p_value

f_value,p_value=variance_ratio_test(array1,array2)
print("F_value:",f_value)
print("P_value:",p_value)

F_value: 3.9372817118349177
P_value: 0.03329866458566044


Q2. Given a significance level of 0.05 and the degrees of freedom for the numerator and denominator of an 
F-distribution, write a Python function that returns the critical F-value for a two-tailed test.

In [10]:
from scipy.stats import f
def critical_f_value(alpha,df1,df2):
    alpha=alpha/2
    f_critical=f.ppf(1-alpha,df1,df2)
    
    return f_critical

f_critical=critical_f_value(0.05,5,20)
print(f"The critical F-Value is: {f_critical}")

The critical F-Value is: 3.289055845680407


Q3. Write a Python program that generates random samples from two normal distributions with known 
variances and uses an F-test to determine if the variances are equal. The program should output the F-value, degrees of freedom, and p-value for the test.

In [12]:
import numpy as np
import scipy.stats as stats

sample_size= 50
variance1=2
variance2=3

samples1= np.random.normal(0,np.sqrt(variance1),size=sample_size)
samples2= np.random.normal(0,np.sqrt(variance2),size=sample_size)

f_value,p_value=stats.f_oneway(samples1,samples2)
degree_freedom= (sample_size-1)*2

print(f"F_value: {f_value:.3f}")
print(f"Degrees of freedom: {degree_freedom}")
print(f"P-value: {p_value:.3f}")

F_value: 0.355
Degrees of freedom: 98
P-value: 0.553


Q4.The variances of two populations are known to be 10 and 15. A sample of 12 observations is taken from 
each population. Conduct an F-test at the 5% significance level to determine if the variances are 
significantly different.

In [ ]:
import numpy as np
import scipy.stats as stats

variance1 = 10
variance2 = 15
sample_size = 12

samples1 = np.random.normal(0, np.sqrt(variance1), size=sample_size)
samples2 = np.random.normal(0, np.sqrt(variance2), size=sample_size)

f_value, p_value = stats.f_oneway(samples1, samples2)
degrees_freedom = (sample_size - 1) * 2

print(f"F-value: {f_value:.3f}")
print(f"Degrees of freedom: {degrees_freedom}")
alpha = 0.05
critical_value = stats.f.ppf(1 - alpha, degrees_freedom=sample_size - 1, degrees_freedom=sample_size - 1)
print(f"Critical value: {critical_value:.3f}")
if p_value < alpha:
    print("Reject the null hypothesis. The variances are significantly different.")
else:
    print("Fail to reject the null hypothesis. The variances are not significantly different.")

Q5. A manufacturer claims that the variance of the diameter of a certain product is 0.005. A sample of 25 
products is taken, and the sample variance is found to be 0.006. Conduct an F-test at the 1% significance 
level to determine if the claim is justified.


Q6. Write a Python function that takes in the degrees of freedom for the numerator and denominator of an 
F-distribution and calculates the mean and variance of the distribution. The function should return the 
mean and variance as a tuple.

In [ ]:
import math
df_num=6
df_num=9

def f_distribution_mean_variance(df_num, df_den):
    if df_num <= 2 or df_den <= 2:
        return "The given degrees of freedom should be greater than 2 for both numerator and denominator."

    mean = df_den / (df_den - 2)
    variance = 2 * df_den**2 * (df_num + df_den - 2) / (df_num * (df_den - 2)**2 * (df_den - 4))

    return (mean, variance)


Q7. A random sample of 10 measurements is taken from a normal population with unknown variance. The 
sample variance is found to be 25. Another random sample of 15 measurements is taken from another 
normal population with unknown variance, and the sample variance is found to be 20. Conduct an F-test 
at the 10% significance level to determine if the variances are significantly different.
Ans. n1=10
n2=15
s1^2=25
s2^2=20
F=s1^2/s2^2
df1 = n1 - 1
df2 = n2 - 1

In [19]:
import scipy.stats as stats

# Given sample variances
s1_squared = 25
s2_squared = 20

# Sample sizes
n1 = 10
n2 = 15

# Degrees of freedom
df1 = n1 - 1
df2 = n2 - 1

# Calculate the F-statistic
F_statistic = s1_squared / s2_squared

# Calculate the p-value
p_value = 2 * (1 - stats.f.cdf(F_statistic, df1, df2))

# Significance level
alpha = 0.10

# Compare p-value to the significance level
if p_value < alpha:
    print("Reject the null hypothesis. Variances are significantly different.")
else:
    print("Fail to reject the null hypothesis. No significant difference in variances.")

Fail to reject the null hypothesis. No significant difference in variances.


Q8. The following data represent the waiting times in minutes at two different restaurants on a Saturday 
night: Restaurant A: 24, 25, 28, 23, 22, 20, 27; Restaurant B: 31, 33, 35, 30, 32, 36. Conduct an F-test at the 5% 
significance level to determine if the variances are significantly different.
Ans. 
H0: σ^2_A = σ^2_B 
H1: σ^2_A ≠ σ^2_B
Restaurant A: n = 7 
Restaurant B: n = 6
Restaurant A: s^2_A = 40
Restaurant B: s^2_B = 33
F = (s^2_A / n_A) / (s^2_B / n_B)

F = (40 / 7) / (33 / 6)

F ≈ 1.039

Using an F-distribution table or calculator, we find the critical F-value to be approximately 3.29.
Since our calculated F-statistic (2.38) is less than the critical F-value (3.29), we fail to reject the null hypothesis.

Q9. The following data represent the test scores of two groups of students: Group A: 80, 85, 90, 92, 87, 83; 
Group B: 75, 78, 82, 79, 81, 84. Conduct an F-test at the 1% significance level to determine if the variances 
are significantly different.
Ans. Group A: 80,85,90,92,87,83
Group B: 75,78,82,79,81,84

H0: σ^2_A = σ^2_B
H1: σ^2_A ≠ σ^2_B
A:n=6
B:n=6
Group A: s^2_A = 44
Group B: s^2_B = 11.33
F = (s^2_A / n_A) / (s^2_B / n_B)

F = (44 / 6) / (11.33 / 6)

F ≈ 4.43

Using an F-distribution table or calculator, we find the critical F-value to be approximately 7.13.

Since our calculated F-statistic (4.43) is less than the critical F-value (7.13), we fail toreject the null hypothesis.